### PR TITLE
CORE-1889 Tapis v3 Migration

### PR DESCRIPTION
**Pull request by psarando**
This PR will rename everything from `agave` and `v2` to `tapis` and `v3`, and update endpoint request and response formatting for Tapis v3.

It might be easier to only review all the commits after `CORE-1889 Rename Agave v2 references to Tapis v3` (but feel free to review that one, and the other 2 minor commits before it, on their own).

[The example JSON in our internal de-docs repo](https://gitlab.cyverse.org/core-sw/de-docs/-/tree/master/tapis) might also be helpful.

### CODE REVIEW

File: src/mescal/tapis_de_v3/jobs.clj [hunk 1]
```diff
@@ -4,38 +4,66 @@
...
+  [get-config-val runtime input-name-set output-name-set {:keys [name arg] :as param}]
+  (let [param-type (mp/get-param-type param)
+        v (get-config-val name)
+        v (cond
+            (contains? input-name-set name) (if (= runtime "DOCKER")
+                                              (preprocess-absolute-path-param-value "/TapisInput" v)
+                                              (preprocess-file-param-value v))
+            (contains? output-name-set name) (if (= runtime "DOCKER")
+                                               (preprocess-absolute-path-param-value "/TapisOutput" v)
+                                               (preprocess-file-param-value v))
```
Comment (psarando) [inline src/mescal/tapis_de_v3/jobs.clj:47]: Tapis apps can now run Docker tools, but Tapis will mount inputs inside the container at `/TapisInput`, and outputs in `/TapisOutput`, by default.

File: src/mescal/tapis_de_v3/jobs.clj [hunk 2]
```diff
@@ -4,38 +4,66 @@
...
+            (map? v)                         (:value v)
+            (= param-type "Flag")            (when v arg)
+            :else                            (str v))]
```
Comment (psarando) [inline src/mescal/tapis_de_v3/jobs.clj:50]: All param values must be submitted as strings, now.

File: src/mescal/tapis_de_v3/jobs.clj [hunk 3]
```diff
@@ -78,37 +108,39 @@
...
+              :appVersion        (:version app)
               :archiveOnAppError true
-              :archivePath       (.tapisFilePath tapis (:output_dir submission))
-              :archiveSystem     (.storageSystem tapis)
-              :notifications     (job-notifications (:callbackUrl submission)))
+              :archiveSystemDir  (.tapisFilePath tapis (:output_dir submission))
+              :archiveSystemId   (.storageSystem tapis)
+              :subscriptions     (job-notifications (:callbackUrl submission))
+              :notes             {:appName        (app-listings/get-app-name app)
+                                  :appDescription (app-listings/get-app-description app)})
```
Comment (psarando) [inline src/mescal/tapis_de_v3/jobs.clj:117]: We can add any notes we want in a job submission.
These fields are used in the `format-job-submisison-response` function.

File: src/mescal/tapis_de_v3/jobs.clj [hunk 4]
```diff
@@ -125,19 +157,26 @@
      (assoc (format-job tapis jobs-enabled? app-info-map job)
             :app-disabled (not (app-enabled? statuses jobs-enabled? app-info))))))
 
+(defn- decode-job-history-description
+  [description]
+  (try+
+    (-> description util/decode-json :message)
+    (catch Exception _
+      description)))
```
Comment (psarando) [inline src/mescal/tapis_de_v3/jobs.clj:165]: For some reason, Tapis stores this JSON in the `description` field as a plain string, so we have to decode it to get to the `message`.

File: src/mescal/tapis_v3.clj [hunk 5]
```diff
@@ -80,31 +80,57 @@
...
                  :as            :json
                  :accept        :json
                  :content-type  :json
                  :form-params   body}))))
 
 (defn list-systems
   [base-url token-info-fn timeout page-len]
-  (tapis-get token-info-fn timeout (curl/url base-url "/v3/systems/") {:page-len page-len}))
+  (tapis-get token-info-fn timeout (curl/url base-url "/v3/systems/") {:page-len page-len
+                                                                       :select   "id,enabled,systemType"
```
Comment (psarando) [inline src/mescal/tapis_v3.clj:92]: It appears these are the only system fields we need to reference.

File: src/mescal/tapis_v3.clj [hunk 6]
```diff
@@ -243,4 +280,4 @@
   [base-url token-info-fn timeout storage-system irods-path]
   (when-not (string/blank? irods-path)
     (let [root-dir (get-root-dir base-url token-info-fn timeout storage-system)]
-      (string/replace irods-path (re-pattern (str "\\Q" root-dir)) ""))))
+      (string/replace irods-path (re-pattern (str "^\\Q" root-dir)) ""))))
```
Comment (psarando) [inline src/mescal/tapis_v3.clj:283]: During testing in QA, a path like `/cyverse/home/psarando/analyses/cyverse-qa-test` would become `/home/psarando/analyses/qa-test` without this fix.

File: src/mescal/util.clj [hunk 7]
```diff
@@ -31,7 +31,7 @@
 
 (def ^:private accepted-timestamp-formats
   ["yyyy-MM-dd'T'HH:mm:ssZZ"
-   "yyyy-MM-dd'T'HH:mm:ss.SSSZZ"])
+   "yyyy-MM-dd'T'HH:mm:ss.SSSSSSSSSZZ"])
```
Comment (psarando) [inline src/mescal/util.clj:34]: Tapis v3 timestamps can have 9 decimal digits now, and this format can still parse seconds with only millisecond precision.

File: src/mescal/tapis_v3.clj [hunk 8]
```diff
@@ -48,116 +48,153 @@
...
+
+(defn- format-search-params
+  "Tapis search params should be formatted like
+   `search=(id.in.id-1,id-2)~(tags.in.term)`"
+  [{id-in :id.in ontology-like :ontology.like}]
+  (let [id-search       (format-search-param "id" id-in)
+        ontology-search (format-search-param "tags" ontology-like)]
+    (merge-with (partial string/join "~")
+                id-search
+                ontology-search)))
```
Comment (slr71) [inline src/mescal/tapis_v3.clj:123]: 👍

**Review by slr71 — APPROVED** [review]
I didn't spot any problems! :shipi

File: src/mescal/tapis_de_v3/app_listings.clj [hunk 9]
```diff
@@ -32,7 +32,9 @@
      :can_rate             false
      :can_run              true
      :deleted              false
-     :disabled             (not (and jobs-enabled? (= "UP" (statuses system))))
+     :disabled             (not (and (boolean (:enabled listing))
+                                     jobs-enabled?
+                                     (statuses system)))
```
[the remark this answers is not in the document]
Comment (psarando) [reply src/mescal/tapis_de_v3/app_listings.clj:37]: I can't remember now, so I better double check 😬

[the remark this answers is not in the document]
Comment (psarando) [reply src/mescal/tapis_de_v3/app_listings.clj:37]: Oh, that's right, `mescal.tapis-de-v3/get-system-statuses` does parse the system `statuses` values as boolean values.

**Review by slr71 — APPROVED** [review]
LGTM 👍

**Comment by slr71** [comment]
Ah, I missed it before, but it looks like a rebase is needed. The code looks good, though. 👍

**Comment by psarando** [comment]
Thanks for the review! 👍

**Pull request merged**